File: STACK/Trapping Rain water/brute.py
from typing import List
def trappingRainWater(nums:List[int])->int:
    n=len(nums)
    pre=[0]*n
    pre[0]=nums[0]
    suf=[0]*n
    suf[n-1]=nums[n-1]
    for i in range(1,n):pre[i]=max(pre[i-1],nums[i])
    for i in range(n-2,-1,-1):suf[i]=max(suf[i+1],nums[i])
    t=0
    for i in range(0,n):
        maxi=min(pre[i],suf[i])
        if maxi>nums[i]:t+=maxi-nums[i]
    return t

nums=[3, 0, 1, 0, 4, 0, 2]
ans=trappingRainWater(nums)
print(ans,end=" ")

File: STACK/Largest Area in a Histrogram/Brute.cpp
#include<bits/stdc++.h>
using namespace std;

vector<int>findNse(vector<int>&nums){
    stack<int>s;
    vector<int>ans(nums.size(),-1);
    int n=nums.size();
    for(int i=n-1;i>=0;i--){
        while(!s.empty() && nums[s.top()]>=nums[i])s.pop();
        ans[i]=s.empty()?n:s.top();
        s.push(i);
    }
    return ans;
    }   
    vector<int>findPse(vector<int>&nums){
        stack<int>s;
        vector<int>ans(nums.size(),-1);
        int n=nums.size();
        for(int i=0;i<n;i++){
            while(!s.empty() && nums[s.top()]>nums[i])s.pop();
            ans[i]=s.empty()?-1:s.top();
            s.push(i);
        }
        return ans;
    }

    int getMaxArea(vector<int>&nums) {
        // code here
        vector<int>pse=findPse(nums);
        vector<int>nse=findNse(nums);
        int maxi=INT_MIN;
        for(int i=0;i<nums.size();i++){
            maxi=max(maxi,nums[i]*(nse[i]-pse[i]-1));
        }
        return maxi;
    }
